### PR TITLE
Check checkmate stalemate

### PR DESCRIPTION
**Pull request by jimxlin**
There are some rubocop offenses that I couldn't figure out how to fix.

### CODE REVIEW

File: app/models/king.rb
Comment (MarciaOrr) [inline app/models/king.rb]: change to get_attackers(row, col).any?

File: app/models/king.rb
Comment (MarciaOrr) [inline app/models/king.rb]: change to 
get_attackers(to_row, to_col).any?

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb]: I chose not to use `any?` here because it's not strictly equivalent to `!...empty?`. `any?` roughly translates to "is there at least one non-false and non-nil element in the array?" This means `any?` is generally slower (although this is only noticeable for larger arrays), and it will return false if there is any array of only nil and/or false elements:
```
[nil, nil, false].any? # false
[nil, nil, false].empty? # false
```

In general, I try not to use `any?` if I'm not using it with a block.  In my view, it's worth the slight decrease in readability for this case.

Comment (jacobwgillespie) [reply app/models/king.rb]: If you want a single method for `!something.empty?`, ActiveSupport has added `.present?` ([definition](https://github.com/rails/rails/blob/bbb84a17213d60f25a61ae4a09111d5adac01a05/activesupport/lib/active_support/core_ext/object/blank.rb#L21-L23)).  You can only use it in a Rails project since it's added by ActiveSupport and not built into Ruby, but this is probably a good use-case for it.

Comment (jimxlin) [reply app/models/king.rb]: Good find, I'll change it to that.

File: app/models/queen.rb
Comment (jacobwgillespie) [inline app/models/queen.rb]: I think the `get_path` method is the same in the Queen and Bishop - it can probably be moved to the Piece class even if it's not used by every piece.

File: app/models/bishop.rb
Comment (jacobwgillespie) [inline app/models/bishop.rb]: I think the get_path method is the same in the Queen and Bishop - it can probably be moved to the Piece class even if it's not used by every piece.

File: app/models/bishop.rb
Comment (jacobwgillespie) [inline app/models/bishop.rb]: Inside `path_to_king` it seems there's a little bit of duplication where we're redefining again how a piece can move - would it make sense to do something like this?

```ruby
return nil unless valid_move?(king.row, king.col)
```

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb:8]: Does `move_legal?` need to ensure you can't move yourself into check?  Or is that handled elsewhere?

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb:56]: **Edit:** actually I think this is what `self_check?` is checking so this may already be covered.  I'll leave the diagram below because it looks pretty.  😛 

`self_check?` may need to be integrated with `valid_move?` somehow though...

----

Actually, maybe the `valid_move?` of every piece should check to ensure that the move doesn't place your own king into check.

I believe this scenario would not currently register as checkmate:

screenshot 2017-04-10 17 28 52

The black queen can't actually move to capture the rook because she's blocking the white bishop.  I think just verifying that any piece movement doesn't cause check should fix this.  (and it's a rule anyways, even outside of checkmate)

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb:14]: This might also need some variant of `self_check?` since the king can't move himself into check either.

File: app/models/piece.rb
Comment (jacobwgillespie) [inline app/models/piece.rb:155]: This means that for each call to `self_check?`, there will be several database queries (update piece, create move, update piece, destroy move, possibly update captured piece) so there's a lot of DB traffic which will affect performance and will have a lot of surface area for something to crash in the middle and leave the DB in an inconsistent state.

However, due to our time constraints (hey, it simulates a real-world project!) I think if this is refactored, it should be done in a later PR as that would likely be time-intensive and involve some kind of temporary in-memory representation of the board which would affect pretty much all of the move logic.

So, just wanted to note this so that it's in the GitHub log, but I don't think it needs to be resolved in this PR.

File: app/models/bishop.rb
Comment (jacobwgillespie) [reply app/models/bishop.rb]: Coming back to this, I think that using `valid_move?` is the right approach, but if it ends up that there's unintended side-effects of using `valid_move?` then lets leave the `path_to_king` methods alone.

**Review by jacobwgillespie — COMMENTED** [review]
Looking good!  A few comments...

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb:8]: Self checks are handled before #valid_move. I've added a #finalize_move! that first calls #self_check?, then calls #move_to!.

The alternative of calling #self_check? in #valid_move? broke several tests, and causes a call loop when two kings are within checking distance of each other.

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb:14]: This line is supposed to replace the calling of #self_check? in the super-method.  A call loop may occur when a King piece uses the #self_check? method.

File: app/models/piece.rb
Comment (jimxlin) [reply app/models/piece.rb:155]: Yes, I understood how costly this would be as I was writing it, but I couldn't think of a better way to implement it without overhauling a lot of the code.

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb:56]: Yes, this is the kind of complicated scenario that #self_check? would catch. I can't think of a less costly implementation of #self_check? that could catch this.

As for `self_check?` and `valid_move?`, see my previous comment:
> Self checks are handled after #valid_move. I've added a #finalize_move! that first calls #self_check?, then calls #move_to!.
...

File: app/models/bishop.rb
Comment (jimxlin) [reply app/models/bishop.rb]: No problems when I change it to use `valid_move?`. However, I have to hold off on that for Queen since `Queen#move_legal?` doesn't exist in the master branch yet.

File: app/models/king.rb
Comment (jacobwgillespie) [reply app/models/king.rb:14]: Oh, I see, it's checking `get_attackers` on the destination space. 👍

File: app/models/bishop.rb
Comment (jacobwgillespie) [reply app/models/bishop.rb]: Let me see about merging the Knight and Queen PRs...

Comment (jacobwgillespie) [reply app/models/bishop.rb]: Okay, PRs merged!

**Comment by jimxlin** [comment]
Things are looking good after merging in changes from the master branch (locally). However, I'm getting the following warnings when running rspec:
```
chess-app/spec/models/queen_spec.rb:5: warning: already initialized constant WHITE_ROW
chess-app/spec/models/knight_spec.rb:4: warning: previous definition of WHITE_ROW was here
chess-app/spec/models/queen_spec.rb:6: warning: already initialized constant BLACK_ROW
chess-app/spec/models/knight_spec.rb:6: warning: previous definition of BLACK_ROW was here
```
Is this ok? It seems like rspec is keeping track of constants across files and noticing when they change.

**Comment by jimxlin** [comment]
I'm also getting these rubocop offenses.
```
app/models/king.rb:33:53: C: Pass &:cant_move? as an argument to all? instead of a block.
    !in_check? && valid_moves.empty? && allies.all? { |ally| ally.cant_move? }
                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
app/models/piece.rb:38:3: C: Assignment Branch Condition size for undo_move! is too high. [27.09/15]
  def undo_move!
  ^^^
```
For the first offense, the alternative doesn't work for private methods.
For the second offense, I can't think of a way to break it up into smaller meaningful methods.

**Comment by jacobwgillespie** [comment]
> warning: already initialized constant

That means a constant is being defined more than once, which could be in a loop (probably a `before` block) or it could be a name collision - those names are global.  My guess here is name collision.  Best practices would probably be to translate them from constants to `let()` blocks.

> I'm also getting these rubocop offenses.

For the first one I'd just disable it with a comment then, or make the method public, whichever makes more sense.

For the second, this might help - not sure if it'll reduce the complexity enough, but possibly:

```ruby
  def undo_move!
    last_move = moves.last
    from_row, from_col = last_move.from_position
    last_move_type, captured_piece_id = last_move.move_type.split

    undo_capture!(captured_piece_id.to_i, row, col) if last_move_type == 'capture'
    last_move.destroy
    update_attributes(row: from_row, col: from_col)
  end
```

If that doesn't do it, I'd add a `is_a_capture?` method to the Move and use it for that if condition.

**Review by jacobwgillespie — APPROVED** [review]
🎉

**Pull request merged**
